src/GraphQL.MicrosoftDI/SelfActivatingServiceProvider.cs

```diff
@@ -37,6 +37,19 @@ public SelfActivatingServiceProvider(IServiceProvider serviceProvider)
         }
 
         /// <inheritdoc cref="ActivatorUtilities.GetServiceOrCreateInstance(IServiceProvider, Type)"/>
-        public object GetService(Type serviceType) => ActivatorUtilities.GetServiceOrCreateInstance(_serviceProvider, serviceType);
+        public object GetService(Type serviceType)
+        {
+            // if the type is an interface, attempt to retrieve the interface registration from the
+            // underlying service provider or else return null. (May trigger an exception in a method
+            // calling GetRequiredService, of course.) But for concrete classes, attempt to
+            // create the instance if it is not able to be returned from the service provider.
+            // Note: this does not intelligently choose the constructor but rather tries the first one it encounters.
+            if (serviceType == typeof(IServiceProvider))
```

Comment:
`if (serviceType == typeof(IServiceProvider) || serviceType == typeof(SelfActivatingServiceProvider))` to be precise.

Reply:
Right. Or if it is assignable - in case of inheritance.

Reply:
Well, in this case it's not sealed, but nor is GetService virtual, so I just did the code you suggested.

Reply:
https://github.com/graphql-dotnet/graphql-dotnet/pull/2769